File: FileSystemTree.java
import java.util.*;

public class FileSystemTree {

    // 1. The Node Structure (N-ary Tree)
    public static class Node {
        String id;
        String name;
        String type; // "folder" or "file"
        String parentId;
        String path;
        List<String> children; // Stores IDs of child nodes

        public Node(String id, String name, String type, String parentId, String path) {
            this.id = id;
            this.name = name;
            this.type = type;
            this.parentId = parentId;
            this.path = path;
            this.children = new ArrayList<>();
        }
    }

    // 2. The HashMaps for O(1) Optimization
    private Map<String, Node> nodes;            // Quick lookup by ID
    private Map<String, Node> pathMap;          // O(1) lookup by full path
    private Map<String, List<Node>> nameMap;    // O(1) lookup by name (handles duplicates)

    public FileSystemTree() {
        this.nodes = new HashMap<>();
        this.pathMap = new HashMap<>();
        this.nameMap = new HashMap<>();
        initRoot();
    }

    private void initRoot() {
        Node root = new Node("root", "Root", "folder", null, "/Root");
        registerNode(root);
    }

    // Helper to keep HashMaps in sync
    private void registerNode(Node node) {
        nodes.put(node.id, node);
        pathMap.put(node.path, node);

        // Add to name map (handling multiple files with same name)
        nameMap.putIfAbsent(node.name, new ArrayList<>());
        nameMap.get(node.name).add(node);
    }

    private void unregisterNode(Node node) {
        nodes.remove(node.id);
        pathMap.remove(node.path);

        if (nameMap.containsKey(node.name)) {
            nameMap.get(node.name).removeIf(n -> n.id.equals(node.id));
            if (nameMap.get(node.name).isEmpty()) {
                nameMap.remove(node.name);
            }
        }
    }

    // ── Public API ────────────────────────────

    // INSERT: O(1) time
    public void insert(String parentId, String name, String type) throws Exception {
        Node parent = nodes.get(parentId);
        if (parent == null) throw new Exception("Parent not found");
        if (!parent.type.equals("folder")) throw new Exception("Cannot add children to a file");

        String id = UUID.randomUUID().toString();
        String path = parent.path + "/" + name;

        if (pathMap.containsKey(path)) throw new Exception("Node already exists at path");

        Node newNode = new Node(id, name, type, parentId, path);
        parent.children.add(id);
        registerNode(newNode);
    }

    // DELETE: O(k) where k is size of subtree
    public void deleteNode(String id) throws Exception {
        if (id.equals("root")) throw new Exception("Cannot delete root");
        Node node = nodes.get(id);
        if (node == null) throw new Exception("Node not found");

        // Remove from parent's children list
        Node parent = nodes.get(node.parentId);
        if (parent != null) {
            parent.children.remove(id);
        }

        // Recursively delete all children
        deleteSubtree(id);
    }

    private void deleteSubtree(String id) {
        Node node = nodes.get(id);
        if (node == null) return;

        // Copy list to avoid ConcurrentModificationException
        List<String> childrenCopy = new ArrayList<>(node.children);
        for (String childId : childrenCopy) {
            deleteSubtree(childId);
        }
        unregisterNode(node);
    }

    // SEARCH 1: O(1) HashMap lookup by Path
    public Node searchByPath(String path) {
        return pathMap.get(path);
    }

    // SEARCH 2: O(1) HashMap lookup by Name
    public List<Node> searchByName(String name) {
        return nameMap.getOrDefault(name, new ArrayList<>());
    }

    // SEARCH 3: O(n) DFS Traversal Fallback
    public List<Node> searchDFS(String query) {
        List<Node> results = new ArrayList<>();
        dfs("root", query.toLowerCase(), results);
        return results;
    }

    private void dfs(String nodeId, String query, List<Node> results) {
        Node node = nodes.get(nodeId);
        if (node == null) return;

        if (node.name.toLowerCase().contains(query)) {
            results.add(node);
        }

        for (String childId : node.children) {
            dfs(childId, query, results);
        }
    }
}

File: backend/fileSystemTree.js
const { v4: uuidv4 } = require('uuid');

// ─────────────────────────────────────────────
//  FileSystemTree – the main in-memory engine
//  Uses: Tree + Path HashMap (O1) + Name HashMap (O1)
// ─────────────────────────────────────────────
class FileSystemTree {
  constructor() {
    // Map<id, node>
    this.nodes = {};
    // HashMap: path → node   (O(1) lookup)
    this.pathMap = {};
    // HashMap: name → [node] (handles duplicates)
    this.nameMap = {};

    this._initRoot();
  }

  _initRoot() {
    const root = {
      id: 'root',
      name: 'Root',
      type: 'folder',
      parentId: null,
      path: '/Root',
      children: [],
    };
    this._registerNode(root);
  }

  // ── Internal helpers ──────────────────────

  _registerNode(node) {
    this.nodes[node.id] = node;
    this.pathMap[node.path] = node;

    if (!this.nameMap[node.name]) this.nameMap[node.name] = [];
    if (!this.nameMap[node.name].find(n => n.id === node.id)) {
      this.nameMap[node.name].push(node);
    }
  }

  _unregisterNode(node) {
    delete this.nodes[node.id];
    delete this.pathMap[node.path];

    if (this.nameMap[node.name]) {
      this.nameMap[node.name] = this.nameMap[node.name].filter(n => n.id !== node.id);
      if (this.nameMap[node.name].length === 0) delete this.nameMap[node.name];
    }
  }

  // ── Public API ────────────────────────────

  getTree() {
    return this.nodes['root'];
  }

  getFullTree() {
    const build = (nodeId) => {
      const node = this.nodes[nodeId];
      if (!node) return null;
      return {
        ...node,
        children: node.children.map(cid => build(cid)).filter(Boolean),
      };
    };
    return build('root');
  }

  insert(parentId, name, type) {
    const parent = this.nodes[parentId];
    if (!parent) throw new Error(`Parent not found: ${parentId}`);
    if (parent.type !== 'folder') throw new Error('Cannot add children to a file');

    const id = uuidv4();
    const path = parent.path + '/' + name;

    if (this.pathMap[path]) throw new Error(`Node already exists at path: ${path}`);

    const node = { id, name, type, parentId, path, children: [] };
    parent.children.push(id);
    this._registerNode(node);
    return node;
  }

  delete(nodeId) {
    const node = this.nodes[nodeId];
    if (!node) throw new Error(`Node not found: ${nodeId}`);
    if (nodeId === 'root') throw new Error('Cannot delete root');

    // Recursively collect all descendants
    const toDelete = [];
    const collect = (id) => {
      const n = this.nodes[id];
      if (!n) return;
      toDelete.push(n);
      n.children.forEach(collect);
    };
    collect(nodeId);

    // Unregister all
    toDelete.forEach(n => this._unregisterNode(n));

    // Remove from parent's children list
    const parent = this.nodes[node.parentId];
    if (parent) parent.children = parent.children.filter(id => id !== nodeId);

    return toDelete.length;
  }

  // ── Search methods ────────────────────────

  searchByName(name) {
    return this.nameMap[name] || [];
  }

  searchByPath(path) {
    return this.pathMap[path] || null;
  }

  // DFS search (fallback, O(n))
  dfsSearch(name, nodeId = 'root') {
    const node = this.nodes[nodeId];
    if (!node) return [];
    const results = [];
    if (node.name.toLowerCase().includes(name.toLowerCase())) results.push(node);
    for (const cid of node.children) {
      results.push(...this.dfsSearch(name, cid));
    }
    return results;
  }

  // Rename a node and recursively update all descendant paths
  rename(nodeId, newName) {
    const node = this.nodes[nodeId];
    if (!node) throw new Error(`Node not found: ${nodeId}`);
    if (nodeId === 'root') throw new Error('Cannot rename root');

    const parent = this.nodes[node.parentId];
    const newPath = parent.path + '/' + newName;
    if (this.pathMap[newPath]) throw new Error(`A node already exists at: ${newPath}`);

    // Collect all nodes under this one (BFS)
    const queue = [node];
    while (queue.length) {
      const cur = queue.shift();
      const oldPath = cur.path;

      // Unregister old
      delete this.pathMap[oldPath];
      if (this.nameMap[cur.name]) {
        this.nameMap[cur.name] = this.nameMap[cur.name].filter(n => n.id !== cur.id);
        if (this.nameMap[cur.name].length === 0) delete this.nameMap[cur.name];
      }

      // Update path/name
      if (cur.id === nodeId) {
        cur.path = newPath;
        cur.name = newName;
      } else {
        cur.path = cur.path.replace(node.path, newPath);
      }

      // Re-register new
      this.pathMap[cur.path] = cur;
      if (!this.nameMap[cur.name]) this.nameMap[cur.name] = [];
      this.nameMap[cur.name].push(cur);

      cur.children.forEach(cid => queue.push(this.nodes[cid]));
    }

    return node;
  }

  // ── Serialization ─────────────────────────

  toJSON() {
    return Object.values(this.nodes);
  }

  loadFromJSON(nodeArray) {
    this.nodes = {};
    this.pathMap = {};
    this.nameMap = {};

    for (const node of nodeArray) {
      this.nodes[node.id] = node;
      this.pathMap[node.path] = node;
      if (!this.nameMap[node.name]) this.nameMap[node.name] = [];
      this.nameMap[node.name].push(node);
    }
  }
}

module.exports = FileSystemTree;
